File: javascript/palindrome-products/palindrome-products.js
const isPalindrome = (num) => [...num.toString()].reverse().join('') === num.toString();

export class Palindromes {
    constructor(smallest, largest) {
        this.smallest = smallest;
        this.largest = largest;
    }

    static generate({ minFactor, maxFactor }) {
        if (minFactor > maxFactor) {
            throw new Error('min must be <= max');
        }

        let smallest = { value: null, factors: [] };
        let largest = { value: null, factors: [] };
        for (let a = minFactor; a <= maxFactor; a++) {
            for (let b = a; b <= maxFactor; b++) {
                const candidate = { value: a * b, factors: [[a, b]] };
                if (isPalindrome(candidate.value)) {
                    if (smallest.value === null || candidate.value < smallest.value) {
                        smallest = candidate;
                    } else if (candidate.value === smallest.value) {
                        smallest.factors.push(candidate.factors[0]);
                    }
                    if (largest.value === null || candidate.value > largest.value) {
                        largest = candidate;
                    } else if (candidate.value === largest.value) {
                        largest.factors.push(candidate.factors[0]);
                    }
                }
            }
        }
        return new Palindromes(smallest, largest);
    }
}

File: ruby/palindrome-products/palindrome_products.rb
class Palindromes
  Palindrome = Struct.new(:value, :factors)

  attr_reader :smallest, :largest

  def initialize(min_factor: 1, max_factor:)
    @min_factor = min_factor
    @max_factor = max_factor
  end

  def generate
    (@min_factor..@max_factor).to_a.repeated_combination(2).each do |a, b|
      product = a * b
      if palindrome? product
        palindrome = Palindrome.new(product, [[a, b]])
        if @smallest.nil? || palindrome.value < @smallest.value
          @smallest = palindrome
        elsif @smallest.value == palindrome.value
          @smallest.factors << palindrome.factors.first
        end

        if @largest.nil? || palindrome.value > @largest.value
          @largest = palindrome
        elsif @largest.value == product
          @largest.factors << palindrome.factors.first
        end
      end
    end
  end

  private

  def palindrome?(n)
    n.to_s.reverse == n.to_s
  end
end
